src/Splunk.OpenTelemetry.AutoInstrumentation/ServiceNameWarning.cs

```diff
@@ -0,0 +1,80 @@
+﻿// <copyright file="ServiceNameWarning.cs" company="Splunk Inc.">
+// Copyright Splunk Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Splunk.OpenTelemetry.AutoInstrumentation.Logging;
+
+namespace Splunk.OpenTelemetry.AutoInstrumentation;
+
+internal static class ServiceNameWarning
+{
+    private const string ServiceNameEnvVarKey = "OTEL_SERVICE_NAME";
+    private const string ResourcesEnvVarKey = "OTEL_RESOURCE_ATTRIBUTES";
+    private const string AttributeName = "service.name";
+    private const char AttributeListSplitter = ',';
+    private const char AttributeKeyValueSplitter = '=';
+
+    private static bool _warningEmitted;
+
+    internal static void SendOnMissingServiceName(ILogger logger)
+    {
+        if (_warningEmitted)
+        {
+            return;
```

Comment:
It is not thread safe.
What is more, it is not necessary to make check 3 times. It will be enough to handle it once.

I think you could use https://learn.microsoft.com/en-us/dotnet/api/system.threading.interlocked.exchange?view=net-6.0#system-threading-interlocked-exchange(system-uint32@-system-uint32) here.